[devel] Errors logged to console

##### Tribler version/branch+revision:

9f1d4d94d1a2df32b42a57017d1246d568821876 devel
##### Relevant log file output:

For all/most of downloads:

```
ERROR   1472497939.37  LaunchManyCore:637   Exception while checkpointing: tails-i386-2.3
Traceback (most recent call last):
  File "/home/vi/home/tribler/tribler/Tribler/Core/APIImplementation/LaunchManyCore.py", line 634, in network_checkpoint_callback
    self.save_download_pstate(infohash, pstate)
  File "/home/vi/home/tribler/tribler/Tribler/Core/APIImplementation/LaunchManyCore.py", line 808, in save_download_pstate
    self.downloads[infohash].save_resume_data())
  File "/home/vi/home/tribler/tribler/Tribler/dispersy/taskmanager.py", line 41, in register_task
    assert not self.is_pending_task_active(name), name
AssertionError: save_pstate 1862.259269
```

--- a/Tribler/Core/APIImplementation/LaunchManyCore.py
+++ b/Tribler/Core/APIImplementation/LaunchManyCore.py
@@ -13,7 +13,7 @@ from threading import Event, enumerate as enumerate_threads
 from traceback import print_exc
 
 from twisted.internet import reactor
-from twisted.internet.defer import Deferred, inlineCallbacks
+from twisted.internet.defer import Deferred, inlineCallbacks, DeferredList
 from twisted.internet.task import deferLater, LoopingCall
 from twisted.internet.threads import deferToThread
 from twisted.python.threadable import isInIOThread
@@ -368,7 +368,7 @@ class TriblerLaunchMany(TaskManager):
             self.downloads[infohash] = d
             setup_deferred = d.setup(dscfg, pstate, initialdlstatus, wrapperDelay=setupDelay,
                                      share_mode=share_mode, checkpoint_disabled=checkpoint_disabled)
-            setup_deferred.addCallback(self.on_download_wrapper_created)
+            setup_deferred.addCallback(self.on_download_handle_created)
 
         if d and not hidden and self.session.get_megacache():
             @forceDBThread
@@ -389,15 +389,12 @@ class TriblerLaunchMany(TaskManager):
 
         return d
 
-    def on_download_wrapper_created(self, (d, pstate)):
-        """ Called by network thread """
-        try:
-            if pstate is None and not d.get_checkpoint_disabled():
-                # Checkpoint at startup
-                (infohash, pstate) = d.network_checkpoint()
-                self.save_download_pstate(infohash, pstate)
-        except:
-            print_exc()
+    def on_download_handle_created(self, download):
+        """
+        This method is called when the download handle has been created.
+        Immediately checkpoint the download and write the resume data.
+        """
+        return download.checkpoint()
 
     def remove(self, d, removecontent=False, removestate=True, hidden=False):
         """ Called by any thread """
@@ -570,7 +567,7 @@ class TriblerLaunchMany(TaskManager):
 
         self.previous_active_downloads = new_active_downloads
         if do_checkpoint:
-            self.session.checkpoint()
+            self.session.checkpoint_downloads()
 
         if self.state_cb_count % 4 == 0 and self.tunnel_community:
             self.tunnel_community.monitor_downloads(states_list)
@@ -668,49 +665,27 @@ class TriblerLaunchMany(TaskManager):
         else:
             self._logger.info("tlm: could not resume checkpoint %s %s %s", filename, tdef, dscfg)
 
-    def checkpoint(self, stop=False, checkpoint=True, gracetime=2.0):
-        """ Called by any thread, assume sesslock already held """
-        # Even if the list of Downloads changes in the mean time this is
-        # no problem. For removals, dllist will still hold a pointer to the
-        # Download, and additions are no problem (just won't be included
-        # in list of states returned via callback.
-        #
-        dllist = self.downloads.values()
-        self._logger.debug("tlm: checkpointing %s stopping %s", len(dllist), stop)
+    def checkpoint_downloads(self):
+        """
+        Checkpoints all running downloads in Tribler.
+        Even if the list of Downloads changes in the mean time this is no problem.
+        For removals, dllist will still hold a pointer to the download, and additions are no problem
+        (just won't be included in list of states returned via callback).
+        """
+        downloads = self.downloads.values()
+        deferred_list = []
+        self._logger.debug("tlm: checkpointing %s downloads", len(downloads))
+        for download in downloads:
+            deferred_list.append(download.checkpoint())
 
-        return deferLater(reactor, 0, self.network_checkpoint_callback, dllist, stop, checkpoint, gracetime)
+        return DeferredList(deferred_list)
 
-    def network_checkpoint_callback(self, dllist, stop, checkpoint, gracetime):
-        """ Called by network thread """
-        if checkpoint:
-            for d in dllist:
-                try:
-                    # Tell all downloads to stop, and save their persistent state
-                    # in a infohash -> pstate dict which is then passed to the user
-                    # for storage.
-                    #
-                    if stop:
-                        (infohash, pstate) = d.network_stop(False, False)
-                    else:
-                        (infohash, pstate) = d.network_checkpoint()
-
-                    self._logger.debug("tlm: network checkpointing: %s %s", d.get_def().get_name(), pstate)
-
-                    self.save_download_pstate(infohash, pstate)
-
-                except Exception as e:
-                    self._logger.exception("Exception while checkpointing: %s", d.get_def().get_name())
-
-        if stop:
-            delay = 0
-            # Some grace time for early shutdown tasks
-            if self.shutdownstarttime is not None:
-                now = timemod.time()
-                diff = now - self.shutdownstarttime
-                if diff < gracetime:
-                    self._logger.info("tlm: shutdown: delaying for early shutdown tasks %s", gracetime - diff)
-                    delay = gracetime - diff
-            return deferLater(reactor, delay, self.network_shutdown)
+    def shutdown_downloads(self):
+        """
+        Shutdown all downloads in Tribler.
+        """
+        for download in self.downloads.values():
+            download.stop()
 
     def remove_pstate(self, infohash):
         def do_remove():

--- a/Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py
+++ b/Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py
@@ -16,7 +16,7 @@ from twisted.internet.task import LoopingCall
 from Tribler.Core import NoDispersyRLock
 from Tribler.Core.DownloadConfig import DownloadStartupConfig, DownloadConfigInterface
 from Tribler.Core.DownloadState import DownloadState
-from Tribler.Core.Libtorrent import checkHandleAndSynchronize, waitForHandleAndSynchronize
+from Tribler.Core.Libtorrent import checkHandleAndSynchronize
 from Tribler.Core.TorrentDef import TorrentDefNoMetainfo, TorrentDef
 from Tribler.Core.Utilities import maketorrent
 from Tribler.Core.Utilities.torrent_utils import get_info_from_handle
@@ -1187,23 +1187,20 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
     #
     # External addresses
     #
-    @waitForHandleAndSynchronize()
     def add_peer(self, addr):
         """ Add a peer address from 3rd source (not tracker, not DHT) to this download.
         @param (hostname_ip,port) tuple
         """
-        self.handle.connect_peer(addr, 0)
+        self.get_handle().addCallback(lambda handle: handle.connect_peer(addr, 0))
 
-    @waitForHandleAndSynchronize()
     def set_priority(self, prio):
-        self.handle.set_priority(prio)
+        self.get_handle().addCallback(lambda handle: handle.set_priority(prio))
 
-    @waitForHandleAndSynchronize(True)
     def dlconfig_changed_callback(self, section, name, new_value, old_value):
         if section == 'downloadconfig' and name == 'max_upload_rate':
-            self.handle.set_upload_limit(int(new_value * 1024))
+            self.get_handle().addCallback(lambda handle: handle.set_upload_limit(int(new_value * 1024)))
         elif section == 'downloadconfig' and name == 'max_download_rate':
-            self.handle.set_download_limit(int(new_value * 1024))
+            self.get_handle().addCallback(lambda handle: handle.set_download_limit(int(new_value * 1024)))
         elif section == 'downloadconfig' and name in ['correctedfilename', 'super_seeder']:
             return False
         return True
@@ -1212,9 +1209,8 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
     def get_share_mode(self):
         return self.handle.status().share_mode
 
-    @waitForHandleAndSynchronize(True)
     def set_share_mode(self, share_mode):
-        self.handle.set_share_mode(share_mode)
+        self.get_handle().addCallback(lambda handle: handle.set_share_mode(share_mode))
 
 
 class LibtorrentStatisticsResponse:

--- a/Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py
+++ b/Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py
@@ -10,7 +10,8 @@ from traceback import print_exc
 
 import libtorrent as lt
 from twisted.internet import defer, reactor
-from twisted.internet.defer import Deferred, CancelledError
+from twisted.internet.defer import Deferred, CancelledError, succeed
+from twisted.internet.task import LoopingCall
 
 from Tribler.Core import NoDispersyRLock
 from Tribler.Core.DownloadConfig import DownloadStartupConfig, DownloadConfigInterface
@@ -19,6 +20,7 @@ from Tribler.Core.Libtorrent import checkHandleAndSynchronize, waitForHandleAndS
 from Tribler.Core.TorrentDef import TorrentDefNoMetainfo, TorrentDef
 from Tribler.Core.Utilities import maketorrent
 from Tribler.Core.Utilities.torrent_utils import get_info_from_handle
+from Tribler.Core.exceptions import SaveResumeDataError
 from Tribler.Core.osutils import fix_filebasename
 from Tribler.Core.simpledefs import (DLSTATUS_WAITING4HASHCHECK, DLSTATUS_HASHCHECKING, DLSTATUS_METADATA,
                                      DLSTATUS_DOWNLOADING, DLSTATUS_SEEDING, DLSTATUS_ALLOCATING_DISKSPACE,
@@ -151,6 +153,9 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
         self._checkpoint_disabled = False
 
         self.deferreds_resume = []
+        self.deferreds_handle = []
+
+        self.handle_check_lc = self.register_task("handle_check", LoopingCall(self.check_handle))
 
     def __str__(self):
         return "LibtorrentDownloadImpl <name: '%s' hops: %d checkpoint_disabled: %d>" % \
@@ -168,6 +173,27 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
     def get_checkpoint_disabled(self):
         return self._checkpoint_disabled
 
+    def check_handle(self):
+        """
+        Check whether the handle exists and is valid. If so, stop the looping call and fire the deferreds waiting
+        for the handle.
+        """
+        if self.handle and self.handle.is_valid():
+            self.handle_check_lc.stop()
+            for deferred in self.deferreds_handle:
+                deferred.callback(self.handle)
+
+    def get_handle(self):
+        """
+        Returns a deferred that fires with a valid libtorrent download handle.
+        """
+        if self.handle and self.handle.is_valid():
+            return succeed(self.handle)
+
+        deferred = Deferred()
+        self.deferreds_handle.append(deferred)
+        return deferred
+
     def setup(self, dcfg=None, pstate=None, initialdlstatus=None,
               wrapperDelay=0, share_mode=False, checkpoint_disabled=False):
         """
@@ -179,6 +205,7 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
         network_create_engine_wrapper.
         """
         # Called by any thread, assume sessionlock is held
+        self.handle_check_lc.start(1, now=False)
         self.set_checkpoint_disabled(checkpoint_disabled)
 
         try:
@@ -338,7 +365,7 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
             self.cew_scheduled = False
 
             # Return a deferred with the callback already being called
-            return defer.succeed((self, pstate))
+            return defer.succeed(self)
 
     def get_anon_mode(self):
         return self.get_hops() > 0
@@ -496,7 +523,7 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
 
         alert_types = ('tracker_reply_alert', 'tracker_error_alert', 'tracker_warning_alert', 'metadata_received_alert',
                        'file_renamed_alert', 'performance_alert', 'torrent_checked_alert', 'torrent_finished_alert',
-                       'save_resume_data_alert')
+                       'save_resume_data_alert', 'save_resume_data_failed_alert')
 
         if alert_type in alert_types:
             getattr(self, 'on_' + alert_type)(alert)
@@ -505,15 +532,12 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
 
     def on_save_resume_data_alert(self, alert):
         """
-        alert to handle save_resume_data_alert
-        it will assign stored resume data in an attribute,
-            and write it to a file
+        Callback for the alert that contains the resume data of a specific download.
+        This resume data will be written to a file on disk.
         """
         resume_data = alert.resume_data
 
-        if not self.pstate_for_restart:
-            self.pstate_for_restart = self.network_get_persistent_state()
-
+        self.pstate_for_restart = self.get_persistent_download_config()
         self.pstate_for_restart.set('state', 'engineresumedata', resume_data)
         self._logger.debug("%s get resume data %s", hexlify(resume_data['info-hash']), resume_data)
 
@@ -528,6 +552,14 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
         # fire callback for all deferreds_resume
         for deferred_r in self.deferreds_resume:
             deferred_r.callback(resume_data)
+
+        # empties the deferred list
+        self.deferreds_resume = []
+
+    def on_save_resume_data_failed_alert(self, alert):
+        # fire errback for all deferreds_resume
+        for deferred_r in self.deferreds_resume:
+            deferred_r.errback(SaveResumeDataError(alert.msg))
 
         # empties the deferred list
         self.deferreds_resume = []
@@ -796,14 +828,13 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
         failure.trap(CancelledError)
         self._logger.error("Resume data cancelled")
 
-    @waitForHandleAndSynchronize()
     def save_resume_data(self):
         """
-        method to save resume data.
+        Save the resume data of a download. This method returns a deferred that fires when the resume data is available.
+        Note that this method only calls save_resume_data once on subsequent calls.
         """
-        # only call libtorrent save resume in the first call
         if not self.deferreds_resume:
-            self.handle.save_resume_data()
+            self.get_handle().addCallback(lambda handle: handle.save_resume_data())
 
         defer_resume = Deferred()
         defer_resume.addErrback(self._on_resume_err)
@@ -1018,7 +1049,7 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
             self._logger.debug("LibtorrentDownloadImpl: network_stop %s", self.tdef.get_name())
             self.cancel_all_pending_tasks()
 
-            pstate = self.network_get_persistent_state()
+            pstate = self.get_persistent_download_config()
             if self.handle is not None:
                 self._logger.debug("LibtorrentDownloadImpl: network_stop: engineresumedata from torrent handle")
                 self.pstate_for_restart = pstate
@@ -1048,11 +1079,8 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
                     self._logger.debug(
                         "LibtorrentDownloadImpl: network_stop: Could not reuse engineresumedata as pstart_for_restart is None")
 
-            # Offload the removal of the dlcheckpoint to another thread
             if removestate:
                 self.session.lm.remove_pstate(self.tdef.get_infohash())
-
-            return (self.tdef.get_infohash(), pstate)
 
     def get_content_dest(self):
         """ Returns the file to which the downloaded content is saved. """
@@ -1078,7 +1106,7 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
                     self.cew_scheduled = True
                     create_engine_wrapper_deferred = self.network_create_engine_wrapper(
                         self.pstate_for_restart, initialdlstatus, share_mode=self.get_share_mode())
-                    create_engine_wrapper_deferred.addCallback(self.session.lm.on_download_wrapper_created)
+                    create_engine_wrapper_deferred.addCallback(self.session.lm.on_download_handle_created)
 
                 can_create_engine_deferred = self.can_create_engine_wrapper()
                 can_create_engine_deferred.addCallback(schedule_create_engine)
@@ -1104,33 +1132,19 @@ class LibtorrentDownloadImpl(DownloadConfigInterface, TaskManager):
         return dest_files
 
     def checkpoint(self):
-        """ Called by any thread """
-        if self._checkpoint_disabled:
-            self._logger.warning("Ignoring checkpoint() call as is checkpointing disabled for this download.")
-        else:
-            infohash, pstate = self.network_checkpoint()
-            reactor.callFromThread(self.session.lm.save_download_pstate, infohash, pstate)
+        """
+        Checkpoint this download. Returns a deferred that fires when the checkpointing is completed.
+        """
+        if self._checkpoint_disabled or not self.handle or not self.handle.is_valid():
+            self._logger.warning("Ignoring checkpoint() call as checkpointing is disabled for this download "
+                                 "or the handle is not ready.")
+            return succeed(None)
 
-    def network_checkpoint(self):
-        """ Called by network thread """
-        with self.dllock:
-            pstate = self.network_get_persistent_state()
-            resdata = None
-            if self.handle is None:
-                if self.pstate_for_restart is not None:
-                    resdata = self.pstate_for_restart.get('state', 'engineresumedata')
-            elif isinstance(self.tdef, TorrentDef):
-                self.save_resume_data()
+        return self.save_resume_data()
 
-            pstate.set('state', 'engineresumedata', resdata)
-            return (self.tdef.get_infohash(), pstate)
-
-    def network_get_persistent_state(self):
-        # Assume sessionlock is held
-
+    def get_persistent_download_config(self):
         pstate = self.dlconfig.copy()
 
-        # Reset unpicklable params
         pstate.set('downloadconfig', 'mode', DLMODE_NORMAL)
 
         # Add state stuff

--- a/Tribler/Core/Libtorrent/LibtorrentMgr.py
+++ b/Tribler/Core/Libtorrent/LibtorrentMgr.py
@@ -57,6 +57,9 @@ class LibtorrentMgr(TaskManager):
         self.metainfo_lock = threading.RLock()
         self.metainfo_cache = {}
 
+        self.process_alerts_lc = self.register_task("process_alerts", LoopingCall(self._task_process_alerts))
+        self.check_reachability_lc = self.register_task("check_reachability", LoopingCall(self._check_reachability))
+
     @blocking_call_on_reactor_thread
     def initialize(self):
         # start upnp
@@ -66,8 +69,8 @@ class LibtorrentMgr(TaskManager):
         self.metadata_tmpdir = tempfile.mkdtemp(suffix=u'tribler_metainfo_tmpdir')
 
         # register tasks
-        self.register_task(u'process_alerts', reactor.callLater(1, self._task_process_alerts))
-        self.register_task(u'check_reachability', reactor.callLater(1, self._task_check_reachability))
+        self.process_alerts_lc.start(1, now=False)
+        self.check_reachability_lc.start(5, now=True)
         self._schedule_next_check(5, DHT_CHECK_RETRIES)
 
         self.register_task(u'task_cleanup_metacache',
@@ -474,14 +477,10 @@ class LibtorrentMgr(TaskManager):
                 for alert in ltsession.pop_alerts():
                     self.process_alert(alert)
 
-        self.register_task(u'process_alerts', reactor.callLater(1, self._task_process_alerts))
-
-    def _task_check_reachability(self):
+    def _check_reachability(self):
         if self.get_session() and self.get_session().status().has_incoming_connections:
-            notify_reachability = lambda: self.notifier.notify(NTFY_REACHABLE, NTFY_INSERT, None, '')
-            self.register_task(u'notify_reachability', reactor.callLater(3, notify_reachability))
-        else:
-            self.register_task(u'check_reachability', reactor.callLater(10, self._task_check_reachability))
+            self.notifier.notify(NTFY_REACHABLE, NTFY_INSERT, None, '')
+            self.check_reachability_lc.stop()
 
     @call_on_reactor_thread
     def _schedule_next_check(self, delay, retries_left):

--- a/Tribler/Core/Libtorrent/__init__.py
+++ b/Tribler/Core/Libtorrent/__init__.py
@@ -1,13 +1,14 @@
 # Written by Egbert Bouman
 
-'''
+"""
 The Libtorrent package contains code to manage the torrent library.
-'''
-import random
-from twisted.internet import reactor
+"""
 
 
 def checkHandleAndSynchronize(default=None):
+    """
+    Return the libtorrent handle if it's available, else return the default value
+    """
     def wrap(f):
         def invoke_func(*args, **kwargs):
             download = args[0]
@@ -16,22 +17,4 @@ def checkHandleAndSynchronize(default=None):
                     return f(*args, **kwargs)
             return default
         return invoke_func
-    return wrap
-
-
-def waitForHandleAndSynchronize(default=None):
-    def wrap(f):
-        def invoke_func(*args, **kwargs):
-            download = args[0]
-            with download.dllock:
-                if download.handle and download.handle.is_valid():
-                    return f(*args, **kwargs)
-                elif not download.session.lm.shutdownstarttime:
-                    lambda_f = lambda a = args, kwa = kwargs: invoke_func(*a, **kwa)
-                    random_id = ''.join(random.choice('0123456789abcdef') for _ in xrange(30))
-                    reactor.callFromThread(
-                        lambda: download.register_task("check_handle_%s" % random_id, reactor.callLater(1, lambda_f)))
-                return default
-        return invoke_func
-
     return wrap

--- a/Tribler/Core/Modules/restapi/settings_endpoint.py
+++ b/Tribler/Core/Modules/restapi/settings_endpoint.py
@@ -95,7 +95,7 @@ class SettingsEndpoint(resource.Resource):
         """
         settings_dict = json.loads(request.content.read())
         self.parse_settings_dict(settings_dict)
-        self.session.save_pstate_sessconfig()
+        self.session.save_session_config()
 
         return json.dumps({"modified": True})
 

--- a/Tribler/Core/Session.py
+++ b/Tribler/Core/Session.py
@@ -176,7 +176,7 @@ class Session(SessionConfigInterface):
         self.notifier = Notifier(use_pool=True)
 
         # Checkpoint startup config
-        self.save_pstate_sessconfig()
+        self.save_session_config()
 
         self.sqlite_db = None
         self.dispersy_member = None
@@ -527,43 +527,42 @@ class Session(SessionConfigInterface):
 
         self.lm.load_checkpoint(initialdlstatus_dict=initialdlstatus_dict)
 
-    def checkpoint(self):
-        """ Saves the internal session state to the Session's state dir. """
-        # Called by any thread
-        self.checkpoint_shutdown(stop=False, checkpoint=True, gracetime=None, hacksessconfcheckpoint=False)
-
     @blocking_call_on_reactor_thread
     def start(self):
-        """ Create the LaunchManyCore instance and start it"""
-
-        # Create engine with network thread
+        """
+        Start a Tribler session by initializing the LaunchManyCore class.
+        Returns a deferred that fires when the Tribler session is ready for use.
+        """
         startup_deferred = self.lm.register(self, self.sesslock)
 
-        if self.get_libtorrent():
-            self.load_checkpoint()
+        def load_checkpoint(_):
+            if self.get_libtorrent():
+                self.load_checkpoint()
 
         self.sessconfig.set_callback(self.lm.sessconfig_changed_callback)
 
-        return startup_deferred
+        return startup_deferred.addCallback(load_checkpoint)
 
     @blocking_call_on_reactor_thread
-    def shutdown(self, checkpoint=True, gracetime=2.0, hacksessconfcheckpoint=True):
-        """ Checkpoints the session and closes it, stopping the download engine.
-        @param checkpoint Whether to checkpoint the Session state on shutdown.
-        @param gracetime Time to allow for graceful shutdown + signoff (seconds).
+    def shutdown(self):
         """
-        # Has to be called from the reactor thread
+        Checkpoints the session and closes it, stopping the download engine.
+        This method has to be called from the reactor thread.
+        """
         assert isInIOThread()
 
         @inlineCallbacks
         def on_early_shutdown_complete(_):
             """
-            Callback that gets called when the early shutdown has been compelted.
+            Callback that gets called when the early shutdown has been completed.
             Continues the shutdown procedure that is dependant on the early shutdown.
             :param _: ignored parameter of the Deferred
             """
-            yield self.checkpoint_shutdown(stop=True, checkpoint=checkpoint,
-                                 gracetime=gracetime, hacksessconfcheckpoint=hacksessconfcheckpoint)
+            self.save_session_config()
+            yield self.checkpoint_downloads()
+            self.lm.shutdown_downloads()
+            self.lm.network_shutdown()
+
             if self.sqlite_db:
                 self.sqlite_db.close()
             self.sqlite_db = None
@@ -652,31 +651,13 @@ class Session(SessionConfigInterface):
     #
     # Internal persistence methods
     #
-    def checkpoint_shutdown(self, stop, checkpoint, gracetime, hacksessconfcheckpoint):
-        """ Checkpoints the Session and optionally shuts down the Session.
-        @param stop Whether to shutdown the Session as well.
-        @param checkpoint Whether to checkpoint at all, or just to stop.
-        @param gracetime Time to allow for graceful shutdown + signoff (seconds).
+    def checkpoint_downloads(self):
         """
-        # Called by any thread
-        with self.sesslock:
-            # Arno: Make checkpoint optional on shutdown. At the moment setting
-            # the config at runtime is not possible (see SessionRuntimeConfig)
-            # so this has little use, and interferes with our way of
-            # changing the startup config, which is to write a new
-            # config to disk that will be read at start up.
-            if hacksessconfcheckpoint:
-                try:
-                    self.save_pstate_sessconfig()
-                except Exception as e:
-                    self._logger.error("save_pstate_sessconfig() failed with error: %s", e)
+        Checkpoints the downloads in Tribler.
+        """
+        return self.lm.checkpoint_downloads()
 
-            # Checkpoint all Downloads and stop NetworkThread
-            if stop:
-                self._logger.debug("Session: checkpoint_shutdown")
-            return self.lm.checkpoint(stop=stop, checkpoint=checkpoint, gracetime=gracetime)
-
-    def save_pstate_sessconfig(self):
+    def save_session_config(self):
         """ Save the runtime SessionConfig to disk """
         # Called by any thread
         sscfg = self.get_current_startup_config_copy()

--- a/Tribler/Core/TorrentChecker/session.py
+++ b/Tribler/Core/TorrentChecker/session.py
@@ -9,7 +9,7 @@ from twisted.internet import reactor, defer
 from twisted.internet.defer import Deferred, maybeDeferred, DeferredList, inlineCallbacks, returnValue
 from twisted.internet.protocol import DatagramProtocol
 from twisted.python.failure import Failure
-from twisted.web.client import Agent, readBody, RedirectAgent
+from twisted.web.client import Agent, readBody, RedirectAgent, HTTPConnectionPool
 
 from Tribler.Core.Utilities.encoding import add_url_params
 from Tribler.Core.Utilities.tracker_utils import parse_tracker_url
@@ -183,6 +183,7 @@ class HttpTrackerSession(TrackerSession):
         self._received_length = None
         self.result_deferred = None
         self.request = None
+        self._connection_pool = HTTPConnectionPool(reactor, False)
 
     def max_retries(self):
         """
@@ -210,7 +211,7 @@ class HttpTrackerSession(TrackerSession):
                               self._announce_page.replace(u'announce', u'scrape')),
                              {"info_hash": self._infohash_list})
 
-        agent = RedirectAgent(Agent(reactor, connectTimeout=self.timeout))
+        agent = RedirectAgent(Agent(reactor, connectTimeout=self.timeout, pool=self._connection_pool))
         self.request = self.register_task("request", agent.request('GET', bytes(url)))
         self.request.addCallback(self.on_response)
         self.request.addErrback(self.on_error)
@@ -320,6 +321,7 @@ class HttpTrackerSession(TrackerSession):
         Cleans the session by cancelling all deferreds and closing sockets.
         :return: A deferred that fires once the cleanup is done.
         """
+        yield self._connection_pool.closeCachedConnections()
         yield super(HttpTrackerSession, self).cleanup()
         self.request = None
 

--- a/Tribler/Core/exceptions.py
+++ b/Tribler/Core/exceptions.py
@@ -72,6 +72,13 @@ class DuplicateTorrentFileError(TriblerException):
     pass
 
 
+class SaveResumeDataError(TriblerException):
+
+    """ This error is used when the resume data of a download fails to save. """
+
+    pass
+
+
 class DuplicateDownloadException(TriblerException):
 
     """ The Download already exists in the Session, i.e., a Download for

--- a/Tribler/Policies/BoostingManager.py
+++ b/Tribler/Policies/BoostingManager.py
@@ -440,7 +440,7 @@ class BoostingManager(TaskManager):
         self.session.set_cm_sources(flag_disabled_sources, CONFIG_KEY_DISABLEDLIST)
         self.session.set_cm_sources(archive_sources, CONFIG_KEY_ARCHIVELIST)
 
-        self.session.save_pstate_sessconfig()
+        self.session.save_session_config()
 
     def log_statistics(self):
         """Log transfer statistics"""

--- a/Tribler/Test/Core/Libtorrent/test_libtorrent_download_impl.py
+++ b/Tribler/Test/Core/Libtorrent/test_libtorrent_download_impl.py
@@ -9,6 +9,7 @@ from Tribler.Core.SessionConfig import SessionStartupConfig
 from Tribler.Core.TorrentDef import TorrentDef
 from Tribler.Core.Utilities.configparser import CallbackConfigParser
 from Tribler.Core.Utilities.twisted_thread import deferred, reactor
+from Tribler.Core.exceptions import SaveResumeDataError
 from Tribler.Core.simpledefs import DLSTATUS_DOWNLOADING, DLMODE_VOD
 from Tribler.Test.Core.base_test import TriblerCoreTest, MockObject
 from Tribler.Test.test_as_server import TestAsServer, TESTS_DATA_DIR
@@ -505,3 +506,21 @@ class TestLibtorrentDownloadImplNoSession(TriblerCoreTest):
 
         self.libtorrent_download_impl.handle.status().pieces = [True * 16]
         self.assertEqual(self.libtorrent_download_impl.get_pieces_base64(), "gA==")
+
+    @deferred(timeout=10)
+    def test_resume_data_failed(self):
+        """
+        Testing whether the correct operations happen when an error is raised during resume data saving
+        """
+        test_deferred = Deferred()
+
+        def on_error(failure):
+            self.assertTrue(failure.check(SaveResumeDataError))
+            test_deferred.callback(None)
+
+        mock_alert = MockObject()
+        mock_alert.msg = "test error"
+
+        self.libtorrent_download_impl.deferreds_resume.append(Deferred().addErrback(on_error))
+        self.libtorrent_download_impl.on_save_resume_data_failed_alert(mock_alert)
+        return test_deferred

--- a/Tribler/Test/Core/Libtorrent/test_libtorrent_download_impl.py
+++ b/Tribler/Test/Core/Libtorrent/test_libtorrent_download_impl.py
@@ -87,8 +87,9 @@ class TestLibtorrentDownloadImpl(TestAsServer):
         tdef = self.create_tdef()
 
         impl = LibtorrentDownloadImpl(self.session, tdef)
-        def callback((ignored, ignored2)):
-            pass
+
+        def callback(_):
+            impl.cancel_all_pending_tasks()
 
         deferred = impl.setup(None, None, None, 0)
         deferred.addCallback(callback)

--- a/Tribler/Test/Core/Libtorrent/test_libtorrent_download_impl.py
+++ b/Tribler/Test/Core/Libtorrent/test_libtorrent_download_impl.py
@@ -1,6 +1,7 @@
 import binascii
 import os
 import libtorrent as lt
+from twisted.internet.defer import Deferred
 
 from Tribler.Core.DownloadConfig import DownloadStartupConfig
 from Tribler.Core.Libtorrent.LibtorrentDownloadImpl import LibtorrentDownloadImpl
@@ -110,6 +111,15 @@ class TestLibtorrentDownloadImpl(TestAsServer):
         impl.dlconfig = DownloadStartupConfig().dlconfig.copy()
         impl.session.lm.on_download_wrapper_created = lambda _: True
         impl.restart()
+
+    @deferred(timeout=20)
+    def test_restart_no_handle(self):
+        test_deferred = Deferred()
+        tdef = self.create_tdef()
+        impl = LibtorrentDownloadImpl(self.session, tdef)
+        impl.session.lm.on_download_handle_created = lambda _: test_deferred.callback(None)
+        impl.restart()
+        return test_deferred
 
     @deferred(timeout=20)
     def test_multifile_torrent(self):

--- a/Tribler/Test/Core/Libtorrent/test_libtorrent_download_impl.py
+++ b/Tribler/Test/Core/Libtorrent/test_libtorrent_download_impl.py
@@ -93,7 +93,7 @@ class TestLibtorrentDownloadImpl(TestAsServer):
 
         deferred = impl.setup(None, None, None, 0)
         deferred.addCallback(callback)
-        return deferred
+        return deferred.addCallback(lambda _: impl.stop())
 
     def test_restart(self):
         tdef = self.create_tdef()
@@ -179,7 +179,7 @@ class TestLibtorrentDownloadImpl(TestAsServer):
         result_deferred = impl.setup(None, None, None, 0)
         result_deferred.addCallback(callback)
 
-        return result_deferred
+        return result_deferred.addCallback(lambda _: impl.stop())
 
 
 class TestLibtorrentDownloadImplNoSession(TriblerCoreTest):

--- a/Tribler/Test/Core/Modules/RestApi/base_api_test.py
+++ b/Tribler/Test/Core/Modules/RestApi/base_api_test.py
@@ -50,8 +50,8 @@ class AbstractBaseApiTest(TestAsServer):
     @blocking_call_on_reactor_thread
     @inlineCallbacks
     def tearDown(self, annotate=True):
-        yield super(AbstractBaseApiTest, self).tearDown(annotate=annotate)
         yield self.close_connections()
+        yield super(AbstractBaseApiTest, self).tearDown(annotate=annotate)
 
     def close_connections(self):
         return self.connection_pool.closeCachedConnections()

--- a/Tribler/Test/Core/Modules/RestApi/base_api_test.py
+++ b/Tribler/Test/Core/Modules/RestApi/base_api_test.py
@@ -6,7 +6,7 @@ from zope.interface import implements
 
 from twisted.internet.defer import succeed, inlineCallbacks
 from twisted.python.threadable import isInIOThread
-from twisted.web.client import Agent, readBody
+from twisted.web.client import Agent, readBody, HTTPConnectionPool
 from twisted.web.http_headers import Headers
 from twisted.web.iweb import IBodyProducer
 
@@ -42,9 +42,19 @@ class AbstractBaseApiTest(TestAsServer):
     @inlineCallbacks
     def setUp(self, autoload_discovery=True):
         yield super(AbstractBaseApiTest, self).setUp(autoload_discovery=autoload_discovery)
+        self.connection_pool = HTTPConnectionPool(reactor, False)
         terms = self.session.lm.category.xxx_filter.xxx_terms
         terms.add("badterm")
         self.session.lm.category.xxx_filter.xxx_terms = terms
+
+    @blocking_call_on_reactor_thread
+    @inlineCallbacks
+    def tearDown(self, annotate=True):
+        yield super(AbstractBaseApiTest, self).tearDown(annotate=annotate)
+        yield self.close_connections()
+
+    def close_connections(self):
+        return self.connection_pool.closeCachedConnections()
 
     @blocking_call_on_reactor_thread
     def setUpPreSession(self):
@@ -59,7 +69,7 @@ class AbstractBaseApiTest(TestAsServer):
         self.config.set_http_api_port(get_random_port(min_port=min_base_port, max_port=min_base_port + 2000))
 
     def do_request(self, endpoint, request_type, post_data, raw_data):
-        agent = Agent(reactor)
+        agent = Agent(reactor, pool=self.connection_pool)
         return agent.request(request_type, 'http://localhost:%s/%s' % (self.session.get_http_api_port(), endpoint),
                              Headers({'User-Agent': ['Tribler ' + version_id]}), POSTDataProducer(post_data, raw_data))
 

--- a/Tribler/Test/Core/Modules/RestApi/test_debug_endpoint.py
+++ b/Tribler/Test/Core/Modules/RestApi/test_debug_endpoint.py
@@ -1,4 +1,5 @@
 import json
+from twisted.internet.defer import inlineCallbacks
 
 from Tribler.Core.Utilities.twisted_thread import deferred
 from Tribler.Test.Core.Modules.RestApi.base_api_test import AbstractApiTest
@@ -13,8 +14,9 @@ from Tribler.dispersy.util import blocking_call_on_reactor_thread
 class TestCircuitDebugEndpoint(AbstractApiTest):
 
     @blocking_call_on_reactor_thread
+    @inlineCallbacks
     def setUp(self, autoload_discovery=True):
-        super(TestCircuitDebugEndpoint, self).setUp(autoload_discovery=autoload_discovery)
+        yield super(TestCircuitDebugEndpoint, self).setUp(autoload_discovery=autoload_discovery)
 
         self.dispersy = Dispersy(ManualEnpoint(0), self.getStateDir())
         self.dispersy._database.open()

--- a/Tribler/Test/Core/Modules/RestApi/test_downloads_endpoint.py
+++ b/Tribler/Test/Core/Modules/RestApi/test_downloads_endpoint.py
@@ -155,9 +155,11 @@ class TestDownloadsEndpoint(AbstractApiTest):
         video_tdef, _ = self.create_local_torrent(os.path.join(TESTS_DATA_DIR, 'video.avi'))
         download = self.session.start_download_from_tdef(video_tdef, DownloadStartupConfig())
         infohash = video_tdef.get_infohash().encode('hex')
+        original_stop = download.stop
 
         def mocked_stop():
             download.should_stop = True
+            download.stop = original_stop
 
         def verify_removed(_):
             self.assertEqual(len(self.session.get_downloads()), 1)

--- a/Tribler/Test/Core/Modules/RestApi/test_multichain_endpoint.py
+++ b/Tribler/Test/Core/Modules/RestApi/test_multichain_endpoint.py
@@ -1,5 +1,6 @@
 import json
 import base64
+from twisted.internet.defer import inlineCallbacks
 
 from Tribler.Core.Utilities.twisted_thread import deferred
 from Tribler.Test.Core.Modules.RestApi.base_api_test import AbstractApiTest
@@ -14,8 +15,9 @@ from Tribler.dispersy.util import blocking_call_on_reactor_thread
 class TestMultichainStatsEndpoint(AbstractApiTest):
 
     @blocking_call_on_reactor_thread
+    @inlineCallbacks
     def setUp(self, autoload_discovery=True):
-        super(TestMultichainStatsEndpoint, self).setUp(autoload_discovery=autoload_discovery)
+        yield super(TestMultichainStatsEndpoint, self).setUp(autoload_discovery=autoload_discovery)
 
         self.dispersy = Dispersy(ManualEnpoint(0), self.getStateDir())
         self.dispersy._database.open()

--- a/Tribler/Test/Core/Modules/RestApi/test_search_endpoint.py
+++ b/Tribler/Test/Core/Modules/RestApi/test_search_endpoint.py
@@ -1,3 +1,4 @@
+from twisted.internet.defer import inlineCallbacks
 from Tribler.dispersy.util import blocking_call_on_reactor_thread
 from Tribler.Core.Utilities.twisted_thread import deferred
 from Tribler.Core.simpledefs import NTFY_CHANNELCAST, NTFY_TORRENTS, SIGNAL_CHANNEL, SIGNAL_ON_SEARCH_RESULTS, \
@@ -32,8 +33,9 @@ class TestSearchEndpoint(AbstractApiTest):
         self.expected_events_messages = []
 
     @blocking_call_on_reactor_thread
+    @inlineCallbacks
     def setUp(self, autoload_discovery=True):
-        super(TestSearchEndpoint, self).setUp(autoload_discovery)
+        yield super(TestSearchEndpoint, self).setUp(autoload_discovery)
         self.channel_db_handler = self.session.open_dbhandler(NTFY_CHANNELCAST)
         self.channel_db_handler._get_my_dispersy_cid = lambda: "myfakedispersyid"
         self.torrent_db_handler = self.session.open_dbhandler(NTFY_TORRENTS)

--- a/Tribler/Test/Core/Video/test_video_server.py
+++ b/Tribler/Test/Core/Video/test_video_server.py
@@ -1,19 +1,75 @@
 # Written by Arno Bakker
 # see LICENSE.txt for license information
 import os
-import time
-import socket
 import binascii
-from traceback import print_exc
+from twisted.internet import reactor
+from twisted.internet.defer import inlineCallbacks, Deferred
+from twisted.internet.endpoints import TCP4ClientEndpoint, connectProtocol
+from twisted.internet.protocol import Protocol, connectionDone
 
 from Tribler.Core.Utilities.network_utils import get_random_port
+from Tribler.Core.Utilities.twisted_thread import deferred
 from Tribler.Core.Video.VideoServer import VideoServer
 from Tribler.Test.Core.base_test import MockObject, TriblerCoreTest
 from Tribler.Test.test_as_server import TESTS_DATA_DIR, TestAsServer
 from Tribler.Core.TorrentDef import TorrentDef
 from Tribler.Core.DownloadConfig import DownloadStartupConfig
+from Tribler.dispersy.util import blocking_call_on_reactor_thread
 
-DEBUG = True
+
+class VideoServerProtocol(Protocol):
+
+    def __init__(self, finished, content_size, expected_content, setset, exp_byte_range):
+        self.finished = finished
+        self.content_size = content_size
+        self.seen_empty_line = False
+        self.has_header = False
+        self.expected_content = expected_content
+        self.setset = setset
+        self.exp_byte_range = exp_byte_range
+
+    def sendMessage(self, msg):
+        self.transport.write("%s" % msg)
+
+    def dataReceived(self, data):
+        if not self.has_header:
+            for line in data.split('\r\n'):
+                if len(line) == 0 and self.seen_empty_line:
+                    self.has_header = True
+                elif len(line) == 0:
+                    self.seen_empty_line = True
+                else:
+                    self.seen_empty_line = False
+                    self.check_header(line)
+        else:
+            assert self.expected_content == data
+            self.transport.loseConnection()
+
+    def connectionLost(self, reason=connectionDone):
+        self.finished.callback(None)
+
+    def check_header(self, line):
+        if line.startswith("HTTP"):
+            if not self.setset:
+                # Python returns "HTTP/1.0 206 Partial Content\r\n" HTTP 1.0???
+                assert line.startswith("HTTP/1.")
+                assert line.find("206") != -1  # Partial content
+            else:
+                assert line.startswith("HTTP/1.")
+                assert line.find("416") != -1  # Requested Range Not Satisfiable
+                self.transport.loseConnection()
+
+        elif line.startswith("Content-Range:"):
+            expline = "Content-Range: bytes " + TestVideoServerSession.create_range_str(
+                self.exp_byte_range[0], self.exp_byte_range[1]) + "/" + str(self.content_size)
+            assert expline == line
+
+        elif line.startswith("Content-Type:") and not self.setset:
+            # We do not check for an exact content-type since that might differ between platforms.
+            assert line.startswith("Content-Type: video")
+
+        elif line.startswith("Content-Length:"):
+            assert line == "Content-Length: " + str(len(self.expected_content))
 
 
 class TestVideoServer(TriblerCoreTest):
@@ -51,14 +107,17 @@ class TestVideoServerSession(TestAsServer):
 
     Mainly HTTP range queries.
     """
-
+    @blocking_call_on_reactor_thread
+    @inlineCallbacks
     def setUp(self, autoload_discovery=True):
         """ unittest test setup code """
-        TestAsServer.setUp(self, autoload_discovery=autoload_discovery)
+        yield TestAsServer.setUp(self, autoload_discovery=autoload_discovery)
         self.port = self.session.get_videoserver_port()
         self.sourcefn = os.path.join(TESTS_DATA_DIR, "video.avi")
         self.sourcesize = os.path.getsize(self.sourcefn)
         self.tdef = None
+        self.expsize = 0
+        yield self.start_vod_download()
 
     def setUpPreSession(self):
         TestAsServer.setUpPreSession(self)
@@ -68,21 +127,22 @@ class TestVideoServerSession(TestAsServer):
     #
     # Tests
     #
+    @deferred(timeout=10)
     def test_specific_range(self):
-        self.range_check(115, 214, self.sourcesize)
+        return self.range_check(115, 214)
 
+    @deferred(timeout=10)
     def test_last_100(self):
-        self.range_check(self.sourcesize - 100, None, self.sourcesize)
+        return self.range_check(self.sourcesize - 100, None)
 
+    @deferred(timeout=10)
     def test_first_100(self):
-        self.range_check(None, 100, self.sourcesize)
+        return self.range_check(None, 100)
 
+    @deferred(timeout=10)
     def test_combined(self):
-        self.range_check(115, 214, self.sourcesize, setset=True)
+        return self.range_check(115, 214, setset=True)
 
-    #
-    # Internal
-    #
     def start_vod_download(self):
         self.tdef = TorrentDef()
         self.tdef.add_content(self.sourcefn)
@@ -93,8 +153,7 @@ class TestVideoServerSession(TestAsServer):
         dscfg.set_dest_dir(os.path.dirname(self.sourcefn))
 
         download = self.session.start_download_from_tdef(self.tdef, dscfg)
-        while not download.handle:
-            time.sleep(1)
+        return download.get_handle()
 
     def get_std_header(self):
         msg = "GET /%s/0 HTTP/1.1\r\n" % binascii.hexlify(self.tdef.get_infohash())
@@ -112,13 +171,7 @@ class TestVideoServerSession(TestAsServer):
 
         return head
 
-    def range_check(self, firstbyte, lastbyte, sourcesize, setset=False):
-        self._logger.debug("range_test: %s %s %s setset %s", firstbyte, lastbyte, sourcesize, setset)
-        self.start_vod_download()
-
-        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        s.connect(('127.0.0.1', self.port))
-
+    def get_header(self, firstbyte, lastbyte, setset=False):
         head = self.get_std_header()
 
         head += "Range: bytes="
@@ -130,102 +183,31 @@ class TestVideoServerSession(TestAsServer):
 
         head += "Connection: close\r\n"
 
-        head += "\r\n"
+        return head + "\r\n"
+
+    def range_check(self, firstbyte, lastbyte, setset=False):
+        test_deferred = Deferred()
+        self._logger.debug("range_test: %s %s %s setset %s", firstbyte, lastbyte, self.sourcesize, setset)
 
         if firstbyte is not None and lastbyte is None:
-            # 100-
-            expfirstbyte = firstbyte
-            explastbyte = self.sourcesize - 1
+            exp_byte_range = (firstbyte, self.sourcesize - 1)
         elif firstbyte is None and lastbyte is not None:
-            # -100
-            expfirstbyte = self.sourcesize - lastbyte
-            explastbyte = self.sourcesize - 1
+            exp_byte_range = (self.sourcesize - lastbyte, self.sourcesize - 1)
         else:
-            expfirstbyte = firstbyte
-            explastbyte = lastbyte
+            exp_byte_range = (firstbyte, lastbyte)
 
         # the amount of bytes actually requested. (Content-length)
-        expsize = explastbyte - expfirstbyte + 1
+        self.expsize = exp_byte_range[1] - exp_byte_range[0] + 1
+        f = open(self.sourcefn, "rb")
+        f.seek(exp_byte_range[0])
 
-        self._logger.debug("Expecting first %s last %s size %s ", expfirstbyte, explastbyte, sourcesize)
-        s.send(head)
+        expdata = f.read(self.expsize)
+        f.close()
 
-        # Parse header
-        s.settimeout(10.0)
-        while True:
-            line = self.read_line(s)
-            if DEBUG:
-                self._logger.debug("Got line: %s", repr(line))
+        def on_connected(p):
+            p.sendMessage(self.get_header(firstbyte, lastbyte, setset))
 
-            if len(line) == 0:
-                if DEBUG:
-                    self._logger.debug("server closed conn")
-                self.assertTrue(False)
-                return
-
-            if line.startswith("HTTP"):
-                if not setset:
-                    # Python returns "HTTP/1.0 206 Partial Content\r\n" HTTP 1.0???
-                    self.assertTrue(line.startswith("HTTP/1."))
-                    self.assertTrue(line.find("206") != -1)  # Partial content
-                else:
-                    self.assertTrue(line.startswith("HTTP/1."))
-                    self.assertTrue(line.find("416") != -1)  # Requested Range Not Satisfiable
-                    return
-
-            elif line.startswith("Content-Range:"):
-                expline = "Content-Range: bytes " + self.create_range_str(
-                    expfirstbyte, explastbyte) + "/" + str(sourcesize) + "\r\n"
-                self.assertEqual(expline, line)
-
-            elif line.startswith("Content-Type:"):
-                # We do not check for an exact content-type since that might differ between platforms.
-                self.assertTrue(line.startswith("Content-Type: video"))
-
-            elif line.startswith("Content-Length:"):
-                self.assertEqual(line, "Content-Length: " + str(expsize) + "\r\n")
-
-            elif line.endswith("\r\n") and len(line) == 2:
-                # End of header
-                break
-
-        data = s.recv(expsize)
-        if len(data) == 0:
-            if DEBUG:
-                self._logger.debug("server closed conn2")
-            self.assertTrue(False)
-            return
-        else:
-            f = open(self.sourcefn, "rb")
-            if firstbyte is not None:
-                f.seek(firstbyte)
-            else:
-                f.seek(lastbyte, os.SEEK_END)
-
-            expdata = f.read(expsize)
-            f.close()
-            self.assertTrue(data, expdata)
-
-            try:
-                # Read body, reading more should EOF (we disabled persist conn)
-                data = s.recv(10240)
-                self.assertTrue(len(data) == 0)
-
-            except socket.timeout:
-                if DEBUG:
-                    self._logger.debug(
-                        "Timeout, video server didn't respond with requested bytes, "
-                        "possibly bug in Python impl of HTTP")
-                    print_exc()
-
-    @staticmethod
-    def read_line(s):
-        line = ''
-        while True:
-            data = s.recv(1)
-            if len(data) == 0:
-                return line
-            else:
-                line = line + data
-            if data == '\n' and len(line) >= 2 and line[-2:] == '\r\n':
-                return line
+        endpoint = TCP4ClientEndpoint(reactor, "localhost", self.port)
+        connectProtocol(endpoint, VideoServerProtocol(test_deferred, self.sourcesize, expdata, setset, exp_byte_range))\
+            .addCallback(on_connected)
+        return test_deferred

--- a/Tribler/Test/test_as_server.py
+++ b/Tribler/Test/test_as_server.py
@@ -85,6 +85,8 @@ class AbstractServer(BaseTestCase):
         from twisted.internet.defer import setDebugging
         setDebugging(True)
 
+    @blocking_call_on_reactor_thread
+    @inlineCallbacks
     def setUp(self, annotate=True):
         self._logger = logging.getLogger(self.__class__.__name__)
 
@@ -95,7 +97,7 @@ class AbstractServer(BaseTestCase):
         defaults.sessdefaults['general']['state_dir'] = self.state_dir
         defaults.dldefaults["downloadconfig"]["saveas"] = self.dest_dir
 
-        self.checkReactor(phase="setUp")
+        yield self.checkReactor(phase="setUp")
 
         self.setUpCleanup()
         os.makedirs(self.session_base_dir)
@@ -158,6 +160,7 @@ class AbstractServer(BaseTestCase):
                                      "Listening ports left on the reactor during %s: %s" % (phase, reader))
 
     @blocking_call_on_reactor_thread
+    @inlineCallbacks
     def tearDown(self, annotate=True):
         self.tearDownCleanup()
         if annotate:
@@ -173,9 +176,9 @@ class AbstractServer(BaseTestCase):
             raise RuntimeError("Couldn't stop the WatchDog")
 
         if self.file_server:
-            return maybeDeferred(self.file_server.stopListening).addCallback(self.checkReactor)
+            yield maybeDeferred(self.file_server.stopListening).addCallback(self.checkReactor)
         else:
-            return self.checkReactor("tearDown")
+            yield self.checkReactor("tearDown")
 
     def tearDownCleanup(self):
         self.setUpCleanup()

--- a/Tribler/Test/test_as_server.py
+++ b/Tribler/Test/test_as_server.py
@@ -3,6 +3,9 @@
 # see LICENSE.txt for license information
 
 # Make sure the in thread reactor is installed.
+from twisted.internet.task import deferLater
+from twisted.internet.tcp import Client
+from twisted.web.http import HTTPChannel
 from Tribler.Core.Utilities.network_utils import get_random_port
 from Tribler.Core.Utilities.twisted_thread import reactor, deferred
 
@@ -131,6 +134,17 @@ class AbstractServer(BaseTestCase):
             for dc in delayed_calls:
                 self._logger.error(">     %s" % dc)
                 dc.cancel()
+
+        has_network_selectables = False
+        for item in reactor.getReaders() + reactor.getWriters():
+            if isinstance(item, HTTPChannel) or isinstance(item, Client):
+                has_network_selectables = True
+                break
+
+        if has_network_selectables:
+            # TODO(Martijn): we wait a while before we continue the check since network selectables
+            # might take some time to cleanup. I'm not sure what's causing this.
+            yield deferLater(reactor, 0.1, lambda: None)
 
         # This is the same check as in the _cleanReactor method of Twisted's Trial
         selectable_strings = []

--- a/twisted/plugins/tribler_plugin.py
+++ b/twisted/plugins/tribler_plugin.py
@@ -66,16 +66,16 @@ class TriblerServiceMaker(object):
         """
         Main method to startup Tribler.
         """
-        def on_shutdown(_):
+        def on_tribler_shutdown(_):
             msg("Tribler shut down")
-            self.process_checker.remove_lock_file()
             reactor.stop()
+            self.process_checker.remove_lock_file()
 
         def signal_handler(sig, _):
             msg("Received shut down signal %s" % sig)
             if not self._stopping:
                 self._stopping = True
-                self.session.shutdown().addCallback(on_shutdown)
+                self.session.shutdown().addCallback(on_tribler_shutdown)
 
         signal.signal(signal.SIGINT, signal_handler)
         signal.signal(signal.SIGTERM, signal_handler)

